pkg/util/hostname/providers.go

```diff
@@ -139,7 +151,22 @@ func saveHostname(cacheHostnameKey string, hostname string, providerName string)
 
 // GetWithProvider returns the hostname for the Agent and the provider that was use to retrieve it
 func GetWithProvider(ctx context.Context) (Data, error) {
-	cacheHostnameKey := cache.BuildAgentKey("hostname")
+	return getHostname(ctx, "hostname", false)
+}
+
+// GetWithProviderWithoutIMDSV2 returns the hostname for the Agent and the provider that was use to retrieve it without using IMDSv2
+func GetWithProviderWithoutIMDSV2(ctx context.Context) (Data, error) {
+	// If the user has set the ec2_prefer_imdsv2 then IMDSv2 is used by default by the user, `legacy_resolution_hostname` is not needed for the transition
+	// If the user has set the ec2_imdsv2_transition then IMDSv2 is used by default by the agent, `legacy_resolution_hostname` is needed for the transition
+	if pkgconfigsetup.Datadog().GetBool("ec2_prefer_imdsv2") || !pkgconfigsetup.Datadog().GetBool("ec2_imdsv2_transition") {
+		return Data{}, nil
+	}
```

Comment:
I think this should be done on the caller side. If we call a method named `GetWithProviderWithoutIMDSV2`, it should return the hostname without IMDSv2. WDYT ?

Reply:
changed the implementation to try to focus the ec2 logic inside the ec2 package